spec/integration/rails/transaction_hooks_spec.rb

```diff
@@ -0,0 +1,406 @@
+          if $raise_on_before_commit[:position]
+            raise 'rollitback_book'
+          end
+        end
+
+        belongs_to :department
+      end
+
+      class EmployeeResource < ApplicationResource
+        self.model = ::Employee
+
+        attribute :first_name, :string
+
+        before_commit only: [:create] do |employee|
+          Callbacks.add(:before_create, employee)
+          if $raise_on_before_commit[:employee]
+            raise 'rollitback'
+          end
+        end
+
+        # Stacking second callback on create
+        before_commit only: [:create] do |employee|
+          Callbacks.add(:stacked_before_create, employee)
+        end
+
+        before_commit only: [:update] do |employee|
+          Callbacks.add(:before_update, employee)
+          if $raise_on_before_commit[:employee]
+            raise 'rollitback'
+          end
+        end
+
+        before_commit only: [:destroy] do |employee|
+          Callbacks.add(:before_destroy, employee)
+          if $raise_on_before_commit[:employee]
+            raise 'rollitback'
+          end
+        end
+
+        after_commit only: :create do |employee|
+          Callbacks.add(:employee_after_create, employee)
+        end
+
+        after_commit do |employee|
+          Callbacks.add(:employee_after_create_eval_test, self)
+        end
+
+        has_many :positions
+      end
+    end
+
+    controller(ApplicationController) do
+      def create
+        employee = resource.build(params)
+
+        if employee.save
+          render jsonapi: employee
+        else
+          raise 'whoops'
+        end
+      end
+
+      def update
+        employee = resource.find(params)
+
+        if employee.update_attributes
+          render_jsonapi(employee, scope: false)
+        else
+          raise 'whoops'
+        end
+      end
+
+      def destroy
+        employee = resource.find(params)
+
+        if employee.destroy
+          render jsonapi: employee
+        else
+          raise 'whoops'
+        end
+      end
+
+      def resource
+        IntegrationHooks::EmployeeResource
+      end
+
+      private
+
+      def params
+        @params ||= begin
+          hash = super.to_unsafe_h.with_indifferent_access
+          hash = hash[:params] if hash.has_key?(:params)
+          hash
+        end
+      end
+    end
+
+    before do
+      @request.headers['Accept'] = Mime[:json]
+      @request.headers['Content-Type'] = Mime[:json].to_s
+
+      routes.draw {
+        post "create" => "anonymous#create"
+        put "update" => "anonymous#update"
+        delete "destroy" => "anonymous#destroy"
+      }
+    end
+
+    def json
+      JSON.parse(response.body)
+    end
+
+    let(:payload) do
+      {
+        data: {
+          type: 'employees',
+          attributes: { first_name: 'Jane' }
+        }
+      }
+    end
+
+    context 'on create' do
+      it 'fires after validations but before ending the transaction' do
```

Comment:
It's implicit this test is talking about `before_commit` only, can we make explicit/?